### ISSUE TITLE
Geting rid of `unstable_setRequestLocale`

### ISSUE THREAD

**Issue by amannn** [amannn/next-intl#663]
As mentioned in [the static rendering docs](https://next-intl-docs.vercel.app/docs/getting-started/app-router#static-rendering), `unstable_setRequestLocale` is intended as a stopgap solution to be removed at some point.

I've started a discussion for the relevant feature on the Next.js side here: https://github.com/vercel/next.js/discussions/58862

If you'd like to help simplify the usage of `next-intl`, please join the discussion there and upvote the comment!

**Comment by xie392** [comment]
![image](https://github.com/amannn/next-intl/assets/61953437/c169d766-38e0-49f5-9541-8f12d268acdb)
Some components need to use "use client", and I want to output static files. There is a conflict between the two. How should I solve it?

**Comment by amannn** [comment]
You don’t need to call the function in client-only pages.

**Comment by A7med3bdulBaset** [comment]
I'm curious about whether calling `setRequestLocale` at the top level `app/[locale]/layout.tsx` will apply to the whole app or not?

When exporting `generateStaticParams` at the layout all sub routes are affected by it.

**Comment by amannn** [comment]
We'll probably introduce support for providing a locale to `next-intl` as an alternative to having to use the middleware to negotiate this (see https://github.com/amannn/next-intl/pull/1017). If your app only supports a single locale, this will enable static rendering without workarounds.

**Comment by amannn** [comment]
A realization while working on https://github.com/amannn/next-intl/pull/1017: We currently support a `defaultLocale` that is used as a fallback when matching locales in the middleware. However, the `defaultLocale` only applies to this particular case.

Developers have previously asked why the `defaultLocale` isn't used when using `next-intl` APIs outside of the `[locale]` folder (see e.g. https://github.com/amannn/next-intl/discussions/1067, but also others).

In case the developer could manage the reading of the locale from params manually in `i18n.ts`, this would also enable returning a fallback locale for cases like this.

Ideally, something like this would be possible:

```tsx
import {notFound} from 'next/navigation';
import {getRequestConfig} from 'next-intl/server';
import {routeParams} from '???';
 
// Can be imported from a shared config
const locales = ['en', 'de'];
 
export default getRequestConfig(async () => {
  let {locale} = routeParams();
  if (!locale) locale = 'en';

  // Validate that the incoming `locale` parameter is valid
  if (!locales.includes(locale as any)) notFound();
 
  return {
    locale,
    messages: (await import(`../messages/${locale}.json`)).default
  };
});
```

**Comment by amannn** [comment]
Hey @vordgi! Nice to hear from you here, been following your work with nimpl-getters! :)

Can you by chance provide a reproduction where you observe that the context is lost? The important part for `setRequestLocale` to work correctly is that it's added to [each and every layout and page](https://next-intl-docs.vercel.app/docs/getting-started/app-router/with-i18n-routing#add-unstable_setrequestlocale-to-all-layouts-and-pages) where you're using APIs from `next-intl`. It should receive the `locale` from `params` as soon as rendering starts and before any APIs from `next-intl` are called.

Are you using any more exotic features from Next.js? E.g. I could imagine that PPR could potentially conflict with this.

Btw. in case you're interested, I had a look at what it'd take for native support from Next.js for providing a "server context" yesterday: https://github.com/vercel/next.js/discussions/58862#discussioncomment-9446295. Leave a comment there if you have additional context!

**Comment by vordgi** [comment]
The parameters have been added everywhere. I will try to create an repro, but basically, it looks like this:

```ts
const Page: React.FC<{ params: { lang: string } }> = async ({ params }) => {
	if (!validLocales.includes(params.lang)) return notFound();

	unstable_setRequestLocale(params.lang);
	const messages = await getMessages();

	return (
		<NextIntlClientProvider messages={messages}>
			{/* Client component with context.provider inside */}
			<ContextProvider value={{key1: [],key2: ''}}>
				<script type="application/ld+json" dangerouslySetInnerHTML={{ __html: JSON.stringify('...')}} />
				{/* Asynchronous server component, translations are only at the top level of this component, below they are lost */}
				<Layout>
					{/* server component whithout translates */}
					<Block1 />
				</Layout>
			</ContextProvider>
		</NextIntlClientProvider>
	);
};
```

Yes, there are indeed risks with PPR. But I think that if functionality is unavailable in part of next.js - just block the functionality in that part. That is, I will add the error "Do not use this in PPR". I hope such an opportunity will be. But in fact, with the current structure of next.js, it is very difficult to find out what the component is inside and it is difficult to rewrite this logic (_I tried_).

I am now looking at the possibility of adding the ability to create a page context. But still only at the beginning of the research

**Comment by vordgi** [comment]
Wish me luck)

https://github.com/vercel/next.js/pull/65897

**Comment by amannn** [comment]
Regarding PPR: On a second thought, maybe this is not an issue at all, since PPR is about pre-rendering as much as possible statically with inner parts remaining dynamic. Dynamic rendering was always the easy part since we can just read a header from the middleware there. Let's see what Next.js 14 brings next week, maybe there are news in regard to PPR and we could do a few first tests.

> getPageContext PR

Oh interesting, thanks for looking into this! As far as I understand the proposal, this would still require providing the locale for every page where you intend to read it, right? In that regard the ergonomics would be similar to `unstable_setRequestLocale`.

I think ideally Next.js should really have a built-in capability to read params deeply in RSC.

**Comment by vordgi** [comment]
In Next.js, you can add parameters to their internal context and then read it the same way I read in `getPathname`. But Next.js stubbornly doesn't want to go that way. So I'm trying to offer them alternatives.

Yes, it will need to be added for each page, it's something comparable to `getStaticProps` before, but you can read from anywhere. In general, this is close to Next.js API and I hope to push some solution.

**Comment by amannn** [comment]
Yep, definitely a good idea to discuss which options we have! I think a tricky part about the story with params is that they can be different per layout/page. A single render might consist of multiple layouts that render and a single page. Theoretically there could be different params at every level (e.g. `/[locale]/[tenant]/page.tsx`).

**Comment by vordgi** [comment]
Since they want to render the layout separately (_this needs to be technically done at first because it works differently now_) - I think in the layout the get-params functionality should be blocked or fixed with params higher in the tree. This is logical in such a situation - it was assembled as a segment once - where it lies.

**Comment by AhmedBaset** [comment]
> I think in the layout the get-params functionality should be blocked or fixed with params higher in the tree.

Currently, you can already read params in the layout via the `params` prop. There would be no need to call `getParams` there. If you did, it should return the same value as the prop.

**Comment by vordgi** [comment]
Decided to make a [PR](https://github.com/vercel/next.js/pull/65922) with the implementation of get-params (_what I talked about above, in the basic version - a very simple change_).

**Comment by vordgi** [comment]
> Currently, you can already read params in the layout via the params prop

Yes, but what about components 10 levels deeper. They now have no idea - they are inside layout or page. And with the current next.js architecture, it's unreal to find out. So I think next.js need to render layout separately and then we can understand what is being rendered and depending on it, return the necessary params or block the API (_although it looks like there shouldnt be any problems with issuance_). But it's still not clear how it would be with PPR, yes...

**Comment by vordgi** [comment]
Hello @amannn. I have to say sorry for incorrect conclusions. The cache works well.

The problem was that I used `setRequestLocale` in `not-found` (_and there defined it dynamically or gave a default_). And this could overwrite what is written at the page level.

_The most annoying thing is that I knew that `<NotFound />` is rendered with the page._

I rewrote it in `not-found` on prop drilling and everything works well.

So, in general, it is always necessary to consider that the `cache`, unlike the `context` (and `AsyncLocalStorage`), is precisely a global storage. And since Layout, NotFound, Error, and Page are rendered as a whole, such situations can arise when there are discrepancies.

**Comment by suitux** [comment]
Any update on getting rid of this?

**Comment by darthmaim** [comment]
How about writing a SWC plugin to automatically inject `unstable_setRequestLocale` into all layouts and pages?

Then people would just have to add it to `experimental.swcPlugins` in the `next.config.js` (or it could be added by `withNextIntl`). The plugin should ignore pages/layouts that use `"use client"` or are not under the route `/[locale]`. The name of the prop should be configurable to support `/[lang]`, `/[language]`, ... and default to `locale` (or maybe  to `['locale', 'lang', 'language']` to work with all 3). The plugin would also have to detect the routes in route groups (`/(test)/[locale]`), parallel routes, intercepting routes, ....

**Comment by amannn** [comment]
Certainly an interesting idea! I don't have experience with writing SWC plugins, but if there's a community effort to build this as a separate package I'd be happy to follow along and potentially explore adding it to the docs or if the approach proves to work well, we could consider adding it to the core. I'm unsure currently if there are edge cases, so this might need so experimentation.

The ideal solution would still require support from the Next.js side I believe: https://github.com/vercel/next.js/discussions/58862.

**Comment by amannn** [comment]
For inspiration, here's an SWC plugin that could be used for inspiration: [`css-variable/swc`](https://github.com/jantimon/css-variable/tree/main/swc) ([usage in Next.js](https://github.com/jantimon/next-swc-plugin-demo/blob/9489ab5902c5a1032ddde0e178f2daac24b22e5b/next.config.js#L4-L7))

**Comment by nevnein** [comment]
This is only tangentially related, but I also feel it does not deserve a separate issue: if you have a root layout outside `[locale]` eg. to render a global non-localized `not-found` [as specified in the docs ](https://next-intl-docs.vercel.app/docs/environments/error-files#catching-non-localized-requests), then Next will opt-out of static rendering on _every_ page, since the root layout does not (and cannot) have the mandatory `unstable_setRequestLocale` call.
I don't think there's a way to solve this without removing the root layout altogether, does anyone have any ideas?

**Comment by WenLonG12345** [comment]
> This is only tangentially related, but I also feel it does not deserve a separate issue: if you have a root layout outside `[locale]` eg. to render a global non-localized `not-found` [as specified in the docs ](https://next-intl-docs.vercel.app/docs/environments/error-files#catching-non-localized-requests), then Next will opt-out of static rendering on _every_ page, since the root layout does not (and cannot) have the mandatory `unstable_setRequestLocale` call. I don't think there's a way to solve this without removing the root layout altogether, does anyone have any ideas?

I am having the same issues with you. Once I remove the layout.tsx and not-found.tsx that outside `[locale]`. `next build` command is successfully built. Any ideas?

**Comment by simplenotezy** [comment]
> You don’t need to call the function in client-only pages.

Does that mean this is not an issue when doing e.g. ISR?

**Comment by ixartz** [comment]
I'm curious about this issue, will it be addressed by Next.js 15 or Next-intl 4?

**Comment by ixartz** [comment]
@amannn Thank you so much for sharing it, I'll read carefully https://next-intl-docs.vercel.app/blog/next-intl-3-22, it seems I totally miss this blog post where you share a lot of information.

**Comment by amannn** [comment]
A note from the discussion at https://github.com/amannn/next-intl/issues/1474: Even if we get rid of `setRequestLocale`, maybe the locale that `getRequestConfig` receives will still stick around. The one rare case seems to be that you want to pass an override for an app that renders content from multiple locales at the same time.

**Comment by amannn** [comment]
Finally some news: https://x.com/sebmarkbage/status/1851728589240115459.

Not much for now, but very happy to hear that _something_  is coming.

**Comment by amannn** [comment]
@reiv The tweet just mentioned that the Next.js team is working on _something_ that will address accessing "global" params like `[locale]` deeply in RSC.

Here's another [mention from Jimmy Lai](https://x.com/feedthejim/status/1852759675881635889):

![Screenshot 2024-11-08 at 15 00 24](https://github.com/user-attachments/assets/5fc6593a-0070-432b-bf9c-c4a21d53bac4)

**Comment by apostolos** [comment]
> It's happening! [vercel/next.js#72837](https://github.com/vercel/next.js/pull/72837) 👀

I'm confused. I didn't expect rootParams to be async. How does this solve https://github.com/vercel/next.js/discussions/71927 ?

Unless... Sebastian mentioned something about being able to suspend above `<html>`. I don't how this will be possible but I guess we'd have to wait and see.

**Comment by amannn** [comment]
Have you seen [the comment below from Sebastian](https://github.com/vercel/next.js/discussions/71927#discussioncomment-11105573)?

**Comment by daveycodez** [comment]
Error: Route /[locale]/about with `dynamic = "error"` couldn't be rendered statically because it used `headers`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering

Even when I have "use client" :/

**Comment by amannn** [comment]
A note for those following along here, it seems like `unstable_rootParams` will be replaced by https://github.com/vercel/next.js/pull/80255 in the future. As of Next.js 15.5, both APIs are available, but `unstable_rootParams` prints a deprecation notice.

While they are mostly equivalent feature-wise, there are some [minor differences](https://github.com/amannn/next-intl/discussions/1627#discussioncomment-14163601).

**Comment by MeesEgberts** [comment]
@amannn `unstable_rootParams` will be removed soon, as mentioned in the [Next.js 16 release blog post](https://nextjs.org/blog/next-16#removals). The post also notes that the team is developing an alternative API, which will be introduced in an upcoming minor release.

**Comment by amannn** [comment]
@MeesEgberts Yes, they're talking about [`next/root-params`](https://github.com/vercel/next.js/pull/80255).

**Comment by ignaciogiri** [comment]
@amannn Can you provide an update on this? Having to set `setRequestLocale` is really affecting the DX like you have no idea. It's extremely painful.

**Comment by amannn** [comment]
> It's extremely painful.

I know -.-

https://github.com/vercel/next.js/pull/80255 that would solve this is already available in the latest Next.js releases (as experimental), but is undocumented for now (see also https://github.com/amannn/next-intl/pull/1632)

I've just reached out to the Next.js team last week with a lengthy message, but didn't get a definitive answer. I saw there's https://github.com/vercel/next.js/pull/90150 though, so maybe that's a sign of life 🤷‍♂️

**Comment by ignaciogiri** [comment]
Yes, I saw, but I can't keep dealing with canary and experimental things. I tried [GT](https://generaltranslation.com/) that offered some support in this matter and then I returned to `next-intl` because of unforeseen bugs.

**Comment by ignaciogiri** [comment]
@amannn I'm so overwhelmed now. No AI can help me bring clarity. Is this relevant or not? https://github.com/vercel/next.js/pull/91189

**Comment by omermecitoglu** [comment]
<details>
  <summary>sorry to ping, i solved the problem already, solution is in the next message</summary>
  
@amannn I tried to remove setRequestLocale by following the tutorial in #1632 but I'm getting this error

<img width="973" height="512" alt="Image" src="https://github.com/user-attachments/assets/1afe5162-a946-4f6e-a08f-500ba456b8ce" />

`/src/i18n/request.ts`

```typescript
import * as rootParams from "next/root-params";
import { hasLocale } from "next-intl";
import { getRequestConfig } from "next-intl/server";
import { notFound, routing } from "./routing";

export default getRequestConfig(async ({ locale }) => {
  if (!locale) {
    const paramValue = await rootParams.locale();
    if (hasLocale(routing.locales, paramValue)) {
      locale = paramValue;
    } else {
      notFound();
    }
  }
  return {
    locale,
    messages: (await import(`../../dictionaries/${locale}.json`)).default,
  };
});
```

`/src/i18n/routing.ts`

```typescript
import { forbidden, notFound, redirect as redirectToOutside, unauthorized } from "next/navigation";
import { createNavigation } from "next-intl/navigation";
import { defineRouting } from "next-intl/routing";

export type AppLocale = typeof routing.locales[number];

export const routing = defineRouting({
  locales: ["en", "tr", "id"],
  defaultLocale: "en",
  localePrefix: {
    mode: "as-needed",
  },
  localeDetection: false,
  pathnames: {
    "/login": "/login",
    "/": "/",
    "/sales": "/sales",
    "/sales/[id]": "/sales/[id]",
    "/scan": "/scan",
  },
});

export const { Link, redirect, usePathname, useRouter, getPathname } = createNavigation(routing);

export { redirectToOutside, notFound, forbidden, unauthorized };
```

Could you tell me what am I doing wrong?

I also checked this example: https://github.com/aurorascharff/next-intl-cache-components but I'm not sure what I am doing different
</details>

**Comment by omermecitoglu** [comment]
I fixed it by separating navigation stuff to `src/i18n/navigation.ts`

**Comment by omermecitoglu** [comment]
🎉

**Comment by amannn** [comment]
Next.js 16.3 is out now with documented, stable support for `next/root-params` 🎉🎉🎉

See the blog post:
→ [**Using next/root-params in Next.js 16.3**](https://next-intl.dev/blog/nextjs-root-params)
